doc/design/topology.md

```diff
@@ -0,0 +1,119 @@
+# Topology Overview
+Topology is a concept in Paddle for representing neural networks.  A neural network contains one topology, which describes how layers connected to each other, and many parameters. The other deep learning frameworks may call this concept a computation graph, neural network configurations.
+
+The topology is not only an API level concept but also how we organize the computation codes for each `Layer` or `Function` in Paddle. The Paddle should maintain a dictionary from `Layer Type` to Layer implementation, e.g.  from string `mul` to function `void tensor_multiply(Tensor& ins, Tensor& outs)'. The mechanism about how to manipulate topology by users, how Paddle maps user topology to implementations of `Layer` and `Function` is a fundamental problem for refactoring Paddle.
+
+## User Stories and Examples
+
+### Kernel Developers
+
+Alan is a professional developer in CPU and GPU. He can write kernel functions of a new `Layer` with the best performance. However, he is not a familiar with Paddle API language, Python. Alan just needs to write the kernel function and register them in Paddle, and then Paddle should generate the user-side APIs for these kernel functions without any codes written by Alan.
+
+```cpp
+template <DeviceType devType>
+void cos_kernel(std::vector<Tensor>& ins, std::vector<Tensor>& outs,  double scale) {
+    // implemetation here.
+}
+
+BEGIN_REGISTER_FUNCTION(cos, cos_kernel)
```

Comment:
Another possible way of achieving this is to make all of these part of Function (i,e. virtual member of Function class).  Beside computation, a Function needs to also implement the interfaces for specifying inputs&outputs, shape inferer, shape checker, estimating flops  etc. 

Need to think which is a better way.